java/flight/flight-core/src/main/java/org/apache/arrow/flight/grpc/RequestContextAdapter.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.flight.grpc;
+
+import java.util.HashMap;
+
+import org.apache.arrow.flight.RequestContext;
+
+import io.grpc.Context;
+
+
+/**
+ * Adapter for holding key value pairs.
+ */
+public class RequestContextAdapter implements RequestContext {
+  private final HashMap<String, String> map = new HashMap<>();
+  public static final Context.Key<RequestContext> REQUEST_CONTEXT_KEY =
+      Context.key("arrow-flight-request-context");
+
+  @Override
+  public String put(String key, String value) {
+    return map.putIfAbsent(key, value);
```

Comment:
This behaviour is surprising (don't update the value if there already is one) given the name of the method.

Reply:
I'll make this throw actually, since we really shouldn't allow overwriting keys.